File: services/api/app/models/checks.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

CheckType = Literal[
    "tests",
    "build",
    "lint",
    "typecheck",
    "coverage",
    "security_sast",
    "dependency_scan",
    "secret_scan",
    "container_scan",
    "accessibility",
    "e2e",
    "custom",
]
CheckSeverity = Literal["info", "warning", "blocking"]
CheckRunTargetType = Literal[
    "project",
    "requirement",
    "epic",
    "dev_task",
    "subtask",
    "pull_request",
    "manual",
]
CheckRunStatus = Literal["pending", "running", "completed", "failed"]
CheckRunConclusion = Literal["success", "failure", "neutral", "skipped", "cancelled"]


class CheckDefinitionCreate(BaseModel):
    code_repository_id: str | None = None
    name: str
    check_type: CheckType
    command: str = ""
    required: bool = True
    enabled: bool = True
    severity: CheckSeverity = "blocking"
    description: str = ""


class CheckDefinitionUpdate(BaseModel):
    name: str | None = None
    check_type: CheckType | None = None
    command: str | None = None
    required: bool | None = None
    enabled: bool | None = None
    severity: CheckSeverity | None = None
    description: str | None = None


class CheckDefinition(BaseModel):
    id: str
    project_id: str
    code_repository_id: str | None = None
    name: str
    check_type: CheckType
    command: str = ""
    required: bool = True
    enabled: bool = True
    severity: CheckSeverity = "blocking"
    description: str = ""
    created_at: datetime
    updated_at: datetime


class CheckDefinitionsFromSafetyProfileRequest(BaseModel):
    code_repository_id: str | None = None


class CheckRunCreate(BaseModel):
    project_id: str
    code_repository_id: str | None = None
    check_definition_id: str | None = None
    target_type: CheckRunTargetType
    target_id: str
    status: CheckRunStatus = "completed"
    conclusion: CheckRunConclusion | None = None
    summary: str = ""
    output: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None


class CheckRun(BaseModel):
    id: str
    project_id: str
    code_repository_id: str | None = None
    check_definition_id: str | None = None
    target_type: CheckRunTargetType
    target_id: str
    status: CheckRunStatus
    conclusion: CheckRunConclusion | None = None
    summary: str = ""
    output: str | None = None
    artifact_id: str | None = None
    command_run_id: str | None = None
    started_at: datetime
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class CheckDefinitionsFromSafetyProfileResponse(BaseModel):
    created: list[CheckDefinition]
    existing: list[CheckDefinition]


class CheckExecutionRequest(BaseModel):
    workspace_id: str
    target_type: CheckRunTargetType = "manual"
    target_id: str | None = None
    timeout_seconds: int | None = None


class CheckExecutionResponse(BaseModel):
    check_run: CheckRun
    command_run: "CommandRun"  # type: ignore[name-defined]


# Late import to resolve forward reference without circular import at module load.
from .commands import CommandRun  # noqa: E402

CheckExecutionResponse.model_rebuild()

File: apps/web/src/types/checks.ts
export type CheckType =
  | 'tests' | 'build' | 'lint' | 'typecheck' | 'coverage'
  | 'security_sast' | 'dependency_scan' | 'secret_scan' | 'container_scan'
  | 'accessibility' | 'e2e' | 'custom'

export type CheckSeverity = 'info' | 'warning' | 'blocking'

export type CheckRunTargetType =
  | 'project' | 'requirement' | 'epic' | 'dev_task' | 'subtask' | 'pull_request' | 'manual'

export type CheckRunStatus = 'pending' | 'running' | 'completed' | 'failed'

export type CheckRunConclusion = 'success' | 'failure' | 'neutral' | 'skipped' | 'cancelled'

export interface CheckDefinitionCreate {
  code_repository_id?: string | null
  name: string
  check_type: CheckType
  command?: string
  required?: boolean
  enabled?: boolean
  severity?: CheckSeverity
  description?: string
}

export interface CheckDefinitionUpdate {
  name?: string
  check_type?: CheckType
  command?: string
  required?: boolean
  enabled?: boolean
  severity?: CheckSeverity
  description?: string
}

export interface CheckDefinition {
  id: string
  project_id: string
  code_repository_id: string | null
  name: string
  check_type: CheckType
  command: string
  required: boolean
  enabled: boolean
  severity: CheckSeverity
  description: string
  created_at: string
  updated_at: string
}

export interface CheckDefinitionsFromSafetyProfileResponse {
  created: CheckDefinition[]
  existing: CheckDefinition[]
}

export interface CheckRunCreate {
  project_id: string
  code_repository_id?: string | null
  check_definition_id?: string | null
  target_type: CheckRunTargetType
  target_id: string
  status?: CheckRunStatus
  conclusion?: CheckRunConclusion | null
  summary?: string
  output?: string | null
  started_at?: string | null
  completed_at?: string | null
}

export interface CheckRun {
  id: string
  project_id: string
  code_repository_id: string | null
  check_definition_id: string | null
  target_type: CheckRunTargetType
  target_id: string
  status: CheckRunStatus
  conclusion: CheckRunConclusion | null
  summary: string
  output: string | null
  artifact_id: string | null
  command_run_id: string | null
  started_at: string
  completed_at: string | null
  created_at: string
  updated_at: string
}

export interface CheckExecutionRequest {
  workspace_id: string
  target_type?: CheckRunTargetType
  target_id?: string | null
  timeout_seconds?: number | null
}

export interface CheckExecutionResponse {
  check_run: CheckRun
  command_run: import('./commands').CommandRun
}
